### DISCUSSION TITLE
Welcome to Hat Labs Discussions!

### DISCUSSION THREAD

**Discussion by mairas** [hatlabs/discussions#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by grahamkendall82** [comment]
Thanks for setting this up Matti! Looking forward to getting my Sailorhat, looks an excellent piece of kit.. Quick question, When compiling a project in platformio, is it just a generic esp32 you choose from the list of devices? ***Edit I just saw its a Wroom32!

**Reply by mairas** [reply]
Yes, I've always had it set up as `esp32dev`. Different ESP32 boards might have slightly different methods for entering the programming mode but SH-ESP32 proudly copies that part from the ESP32 DevKit boards. :-)

**Comment by baden03** [comment]
SH-RPi looks _very_ interesting. Is a specific OS required for the rpi? What about boot options: will it support rpi using external SSD boot?

**Reply by mairas** [reply]
As long as the RPi runs Linux, anything should be fine. The automated installation script is only tested on Raspberry Pi OS.

To allow simultaneous use with other hats reserving the standard SPI pins, the SPI interface used for communicating with the CAN bus controller (NMEA 2000 interface) is on non-standard GPIO pins. Linux kernel device tree overlays are used to create a new spi interface on those pins. I would assume any Linux distribution that supports the Raspberry Pi hardware would also support the device tree overlays.

In addition to the device tree overlay changes, a simple daemon service written in Python is used for communicating with the SH-RPi. The program might have some assumptions specific to Raspberry Pi OS but could be easily fixed.

SH-RPi doesn't interfere with the normal OS boot process or operation, so external SSD booting works just fine!

I received an update from the fabrication house just this morning that the boards are ready and being prepared for delivery. I'd expect to have some initial stock available in a week or so!

**Comment by Mikko-okkiM** [comment]
Terve Matti,

I have 9 meter Finnish walk around boat Marsund 30 with Yanmar 4JH110 Common Rail engine. 4JH110 is a modern diesel engine with CAN bus, but without optional NMEA2000 cables.

I use my own software in Raspberry Pi connected to couple of Wemos D1 Mini boards (via Mosquitto protocol over WLAN) together with different kind of sensors. RPi communicates with boars (using NodeJs) and stores measurements to RPi. Measurements are then monitored by browsers (using my script made with Vue JavaScript framework) at mobile phones and tablets.

I have already purchased couple of CAN transceivers and receivers (MCP2515 TJA1050, MCP2551, SN65HVD230) to be used with ESP32 and ESP8266 boards but I haven't had enough courage to connect them to CAN bus at the boat because they are not isolated.

Is it so that SH-ESP32 CAN pins CANH, CANL and GND can safely be connected directly to corresponding CAN bus wires at the boat? And 12V pin will not be connected at all? 

There seems to be text "CAN term" at the PCB picture of SH-ESP32. Is it so that by soldering little piece of tin to that place then SH-ESP32 acts also as a CAN bus termination resistor?

Mikko Löhönen
Naantali, Finland

**Reply by mairas** [reply]
Moi Mikko,

It's fine to use the CAN interface on SH-ESP32 for generic CAN communications. The electrical characteristics of all CAN buses are the same (as long as they're "classic" CAN and not the faster CAN-FD, but I assume that's the case here).

There's one important detail, though: the interface side of the isolated CAN interface on SH-ESP32 is not internally powered, so you need to provide 12V to it (24V would work as well). When used with N2K, it of course uses the N2K 12V supply. I can think of two options:

1. Connect the GND and 12V to both the SH-ESP32 power input and CAN interface (or add a jumper wire to the wire link pins on board). Then feed those from the same 12V feed as the engine ECU (preferably patching to the wires close to the ECU). This setup does not have galvanic isolation, but that doesn't matter because there are no galvanic connections between the SH-ESP32 and any other devices besides the ECU. So in essence, the isolation barrier is around the whole device.

2. Provide external power and GND to SH-ESP32 power input, and separate GND and 12V from the ECU feed to the CAN interface. You'd have isolation but no real benefit from doing so, other than being able to control the SH-ESP32 power independently of the engine.

3. You could provide external power to SH-ESP32 and the CAN interface, and connect H, L, and GND to the ECU CAN. You'd lose isolation and expose the engine ECU to ground loop noise while having no real benefits, so don't do that. :-)

Finally, you're right about the CAN termination resistor!

**Reply by Mikko-okkiM** [reply]
![sailorhat](https://user-images.githubusercontent.com/80455592/147947605-42eb76da-92ff-4999-8f87-514b6e3db1f2.gif)

I tried to draw a picture of all those three connections you mentioned. I drew the third wire, brown wire, also just in the case that there goes some kind of third wire together with CANH and CANL wires in my boat. If the third wire do exist then should it be connected to CAN GND pin of the SH-ESP32? It would actually mean that it would be connected directly to the ground of the main battery. Or should it be left without any connection?

**Reply by mairas** [reply]
![147947605-42eb76da-92ff-4999-8f87-514b6e3db1f2](https://user-images.githubusercontent.com/407010/148057995-bb3d466c-dc51-4336-adbd-1710326a0f58.png)

Made small modifications. In all cases, +12V and GND (red and blue) should be routed together, and in cases 2 and 3 you'd normally  just have one battery. But personally, I would only consider case number 1.

The third CAN interface wire would be GND, and I assume it does exist. I think you could leave it unconnected because that would remove one potential ground loop.

**Reply by Mikko-okkiM** [reply]
Hello,

for me it seems like the wiring has not changed in case number 1. Perhaps you just wanted to highlight that SH_ESP32 does not need so much amperes than ECU and do not touch wiring between the power source and ECU?

I ordered one SH-ESP32 last night :-)

**Reply by mairas** [reply]
Thanks for the order! It'll be shipped today. :-)

True, wiring didn't change in case 1. What I wanted to highlight is that to avoid inducing noise on the wires, both the positive and negative leads should come from the same source. Drawing style was just due to the art being a ghetto job. :-D

The wires _can_ be much thinner, true, but if you go much thinner than 1.5mm2, mechanical durability might become an issue. Also, if you patch the power with very thin wires, you should add a smaller hanging fuse close to the junction point to protect the cable from shorts.
